Userland/Libraries/LibELF/Arch/x86_64/plt_trampoline.S

```diff
@@ -0,0 +1,58 @@
+/*-
+ * Copyright (c) 1998, 2002 The NetBSD Foundation, Inc.
+ * All rights reserved.
+ *
+ * This code is derived from software contributed to The NetBSD Foundation
+ * by Christos Zoulas and by Charles M. Hannum.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This asm method is copied from NetBSD. We changed the internal method that
+ * gets called and the name, but it's still essentially the same as
+ * _rtld_bind_start from libexec/ld.elf_so/arch/i386/rtld_start.S
+ */
+
+.align	4
+	.globl	_plt_trampoline
+	.hidden	_plt_trampoline
+	.type	_plt_trampoline,@function
+_plt_trampoline:	# (obj, reloff)
+	pushf				# save registers
+	push	%rax
+	push	%rcx
+	push	%rdx
+
+	push	20(%esp)		 # Copy of reloff
+	push	20(%esp)		 # Copy of obj
+	call	_fixup_plt_entry@PLT # Call the binder
```

Comment:
I would verify that this matches the reference asm in netbsd for the x86_64 arch. I wouldn't be surprised if it needs 8 byte alignment for example. FWIW, the 32-bit plt trampoline part is known to not work -- neither itamar nor myself have taken the time to re-enable lazy relocations.

Reply:
I just copied the i386 code and fixed push and pop nothing more so there will probably be some more changes needed....

Reply:
looking at the netbsd implimentation, I have no clue what to change  
Do you have any expertise of this topic?

ref:  
http://ftp.jp.freebsd.org/pub/NetBSD/NetBSD-current/src/libexec/ld.elf_so/arch/x86_64/rtld_start.S

Reply:
So essentially what's happening is this function is "_rtld_bind". In the procedure linkage table (PLT) of each loaded ELF object we have entries for every external symbol we found during the build-time link. The table has something along the lines of a relative jump to a certain point in the global offset table, and a relocation index. The dimension dynamic loader puts this procedure's address in the GOT as a function pointer for all the PLT entries to jump up so that we can call into the dynamic object or dynamic loader for that object (library) to perform symbol lookup and patch the PLT at runtime. 

Which is all called "lazy symbol resolution" and has the purpose of speeding up application load times at the cost of a once per symbol relocation during the execution of the program. Which has security implications and such. The implementation in NetBSD essentially does a manual procedure call prologue/epilogue for all the caller saved registers in the x86_64 ABI. It also does some adjustments to make sure the values that were pushed by the stub code in the PLT are in the proper registers for the C linkage function it ends up calling to do the work of the lazy relocation.

Reply:
You know what? I gave you access to the fork.  
The branch is called Hendiadyoin1:x86_64-Support!  
Feel free to do that part and everything that catches your eye